wremnants/CardTool.py

```diff
@@ -597,6 +598,112 @@ def writeForProcesses(self, syst, processes, label, check_systs=True):
         if syst != self.nominalName:
             self.fillCardWithSyst(syst)
 
+    def setSimultaneousABCD(self):
+        # Having 1 process for fakes, for each bin 3 free floating parameters, 2 normalization for lowMT and highMT and one fakerate between iso and anti iso
+        logger.info(f"Set processes for simultaneous ABCD fit")
+
+        # expected fake contribution
+        hist_fake = sum([group.hists[self.nominalName] if name == "Data" else -1*group.hists[self.nominalName] for name, group in self.datagroups.groups.items()])
+        
+        # setting errors to 0
+        hist_fake.view(flow=True)[...] = np.stack((hist_fake.values(flow=True), np.zeros_like(hist_fake.values(flow=True))), axis=-1)
+
+        if "passIso" not in hist_fake.axes.name or "passMT" not in hist_fake.axes.name:
+            raise RuntimeError('"passIso" and "passMT" expected to be found in histogram, but only have axes ')
+
+        # axes in the correct ordering
+        axes = ["passIso", "passMT"]
+        axes = [*axes, "charge"] if "charge" not in self.project else axes
+        axes += [ax for ax in self.project if ax not in axes]
+
+        if set(hist_fake.axes.name) != set(axes) or hist_fake.axes.name[0] != "passIso" or hist_fake.axes.name[1] != "passMT":
+            logger.debug("Axes in histogram are not the same as required or in a different order than expected, try to project")
+            hist_fake = hist_fake.project(*axes)
+
+        # set the expected values in the signal region
+        hist_fake.view(flow=True)[1,1,...] = sel.fakeHistABCD(hist_fake).view(flow=True)
+        
+        fakename = "Nonprompt"
+
+        self.datagroups.addGroup(fakename, label = "Nonprompt", color = "grey", members=[],)
+        self.datagroups.groups[fakename].hists[f"{self.nominalName}"] = hist_fake
+
+        bin_sizes = [ax.size for ax in hist_fake.axes if ax.name in axes and ax.name not in ["passIso","passMT"]]
+        axes = [ax.name for ax in hist_fake.axes if ax.name in axes and ax.name not in ["passIso","passMT"]]
+
+        for i in range(np.product(bin_sizes)):
+            current_i = i
+            ax_idx = []
+            for num in reversed(bin_sizes):
+                ax_idx.insert(0, current_i % num)
+                current_i //= num
+
+            bin_name = "_".join([f"{ax}{ax_idx[j]}" for j, ax in enumerate(axes)])
+
+            logger.debug(f"Now at {i}/{np.product(bin_sizes)}: {bin_name}")
+
+            other_indices = {ax: ax_idx[j] for j, ax in enumerate(axes)}
+
+            n_failMT_failIso = hist_fake[{**{"passIso":0, "passMT":0}, **other_indices}].value
```

Comment:
It would be nice to have a way to not hardcode the axis names. In #262 I added some features to read the hist in advance, which might be useful, then you can verify that the axes are correct before making assumptions about their names

Reply:
The check that the axis is available is done in L611

Reply:
Ok yes I see. It would still perhaps be better to have iso and mt ax variables rather than always using their names. If you make this into it's own helper class they could be setable variables

Reply:
I define the passIso and passMT now in the common.py since it is also used there, and for example in histselections